crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -390,28 +426,15 @@ impl InstancePool {
             free_list.swap_remove(free_index)
         };
 
-        let host_state = std::mem::replace(&mut req.host_state, Box::new(()));
-
         unsafe {
-            let instance = self.instance(index);
-
-            instance.module = req.module.clone();
-            instance.offsets = VMOffsets::new(
-                std::mem::size_of::<*const u8>() as u8,
-                instance.module.as_ref(),
-            );
-            instance.host_state = host_state;
-
-            Self::set_instance_memories(
-                instance,
-                self.memories.get(index),
-                self.memories.max_wasm_pages,
-            )?;
-            Self::set_instance_tables(instance, self.tables.get(index), self.tables.max_elements)?;
-
-            initialize_vmcontext(instance, req);
-
-            Ok(InstanceHandle::new(instance as _))
+            self.setup_instance(index, req).or_else(|e| {
+                // Deallocate the allocated instance on error
+                let instance = self.instance(index);
+                self.deallocate(&InstanceHandle {
```

Comment:
To confirm, are there tests for this? It seems a bit fishy that we're  deallocating a half-initialized instance without many modifications below. I'd expect, for example `decommit_memory_pages` to fail one one way or another if called on an othewise unmapped region?

Reply:
The `test_pooling_allocator_initial_limits_exceeded` test does cover this, albeit with empty memory/table lists on `deallocate`, as that test is what caught the problem: with an instance limit of 1, it failed to allocate any more instances because it wasn't being added back to the free list.  I will add another memory to the module definition such that the list of initialized memories isn't empty, just in case.

While the instance isn't completely initialized, it should be safe to call `deallocate` on it as we iterate back through only initialized `Memory` and `Table` to call `decommit_memory_pages` on pages that were either made accessible (for `Memory`) or explicitly committed (for `Table`) prior to inserting into those lists.